### ISSUE TITLE
Socket timeout when writing to dynamo

### ISSUE THREAD

**Issue by mariaines** [localstack/localstack#3561]
# Type of request: This is a ...

[x] bug report
[ ] feature request

# Detailed description
We are seeing socket timeouts when writing to dynamo fairly frequently. We are using aiohttp via [aiobotocore](https://github.com/aio-libs/aiobotocore), not sure if that is interacting in some way?

Reads and queries are fine.

## Expected behavior
Dynamo writes succeed within ~seconds

## Actual behavior
Dynamo write times out after several minutes

Stack trace:
```py
ss2_gql_1               | Traceback (most recent call last):
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/graphql/execution/execute.py", line 630, in await_result
ss2_gql_1               |     return await result
ss2_gql_1               |   File "./service/extensions/load_user_extension.py", line 18, in resolve
ss2_gql_1               |     result = await result
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/ariadne/types.py", line 58, in resolve
ss2_gql_1               |     result = await result
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/ariadne/types.py", line 58, in resolve
ss2_gql_1               |     result = await result
ss2_gql_1               |   File "./service/resolvers/loggable_types.py", line 53, in async_log_resolver
ss2_gql_1               |     response = await response
ss2_gql_1               |   File "./service/resolvers/mutation_resolver.py", line 49, in resolve_start_conversation
ss2_gql_1               |     is_closed_to_group,
ss2_gql_1               |   File "./service/controllers/messaging_controller.py", line 74, in create_conversation
ss2_gql_1               |     conversation = await conversation.save(context)
ss2_gql_1               |   File "./service/models/database_model.py", line 674, in save
ss2_gql_1               |     response = await self.get_db_client().update_item(**params)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/aiobotocore/client.py", line 77, in _make_api_call
ss2_gql_1               |     operation_model, request_dict, request_context)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/aiobotocore/client.py", line 96, in _make_request
ss2_gql_1               |     return await self._endpoint.make_request(operation_model, request_dict)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/aiobotocore/endpoint.py", line 74, in _send_request
ss2_gql_1               |     exception):
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/aiobotocore/endpoint.py", line 176, in _needs_retry
ss2_gql_1               |     caught_exception=caught_exception, request_dict=request_dict)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/botocore/hooks.py", line 356, in emit
ss2_gql_1               |     return self._emitter.emit(aliased_event_name, **kwargs)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/botocore/hooks.py", line 228, in emit
ss2_gql_1               |     return self._emit(event_name, kwargs)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/botocore/hooks.py", line 211, in _emit
ss2_gql_1               |     response = handler(**kwargs)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py", line 183, in __call__
ss2_gql_1               |     if self._checker(attempts, response, caught_exception):
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py", line 251, in __call__
ss2_gql_1               |     caught_exception)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py", line 277, in _should_retry
ss2_gql_1               |     return self._checker(attempt_number, response, caught_exception)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py", line 317, in __call__
ss2_gql_1               |     caught_exception)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py", line 223, in __call__
ss2_gql_1               |     attempt_number, caught_exception)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/botocore/retryhandler.py", line 359, in _check_caught_exception
ss2_gql_1               |     raise caught_exception
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/aiobotocore/endpoint.py", line 137, in _do_get_response
ss2_gql_1               |     http_response = await self._send(request)
ss2_gql_1               |   File "/usr/local/lib/python3.7/site-packages/aiobotocore/endpoint.py", line 218, in _send
ss2_gql_1               |     request.method, url=url, headers=headers_, data=data, proxy=proxy)
ss2_gql_1               |   File "/usr/local/lib64/python3.7/site-packages/aiohttp/client.py", line 544, in _request
ss2_gql_1               |     await resp.start(conn)
ss2_gql_1               |   File "/usr/local/lib64/python3.7/site-packages/aiohttp/client_reqrep.py", line 890, in start
ss2_gql_1               |     message, payload = await self._protocol.read()  # type: ignore
ss2_gql_1               |   File "/usr/local/lib64/python3.7/site-packages/aiohttp/streams.py", line 604, in read
ss2_gql_1               |     await self._waiter
ss2_gql_1               | aiohttp.client_exceptions.ServerTimeoutError: Timeout on reading data from socket
```

# Steps to reproduce

## Command used to start LocalStack
docker-compose up

```yml
  localstack:
    # Version 0.12.4 introduces an Elasticsearch domain config issue
    image: localstack/localstack-full:0.12.4
    ports:
      - "4566:4566"  # Edge port
      - "${PORT_WEB_UI-8080}:${PORT_WEB_UI-8080}" # Console UI
      - "5891:5891"
    environment:
      - DOCKER_HOST=unix:///var/run/docker.sock
      - SERVICES=dynamodb,dynamodbstreams,kinesis,lambda,sqs
      - DEFAULT_REGION=us-west-2
      - DEBUG=1
      # Runs Lamba function in lambci container that matches
      # Lamba environment configuration
      - LAMBDA_EXECUTOR=docker-reuse
      - LAMBDA_REMOTE_DOCKER=False
      - LAMBDA_DOCKER_NETWORK=local_services_network
      # If true, will remove idle Lambda containers
      - LAMBDA_REMOVE_CONTAINERS=False
      - AWS_DEFAULT_REGION=us-west-2
      - AWS_CONTAINER_CREDENTIALS_RELATIVE_URI=/creds
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      # Mount the Lambda build directory so Lambdas can be shared across containers
      - /tmp/.lambda-build:/tmp/.lambda-build
    tmpfs:
      - /tmp/localstack:exec,mode=600
    networks:
      services_network:
        ipv4_address: "169.254.170.3"
    depends_on:
      - ecs-local-endpoints
```

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)

```sh
   aws dynamodb create-table \
        --endpoint-url=http://$HOST:4566 \
        --table-name table2 \
        --attribute-definitions \
            AttributeName=_pk,AttributeType=S \
            AttributeName=_sk,AttributeType=S \
            AttributeName=_gsi1_pk,AttributeType=S \
            AttributeName=_gsi1_sk,AttributeType=S \
            AttributeName=_gsi2_pk,AttributeType=S \
            AttributeName=_gsi2_sk,AttributeType=S \
        --key-schema \
            AttributeName=_pk,KeyType=HASH \
            AttributeName=_sk,KeyType=RANGE \
        --provisioned-throughput \
            ReadCapacityUnits=5,WriteCapacityUnits=5 \
        --global-secondary-indexes \
            'IndexName=_gsi1_pk-_gsi1_sk-index,KeySchema=[{AttributeName=_gsi1_pk,KeyType=HASH},{AttributeName=_gsi1_sk,KeyType=RANGE}],Projection={ProjectionType=ALL},ProvisionedThroughput={ReadCapacityUnits=5,WriteCapacityUnits=5}'\
            'IndexName=_gsi2_pk-_gsi2_sk-index,KeySchema=[{AttributeName=_gsi2_pk,KeyType=HASH},{AttributeName=_gsi2_sk,KeyType=RANGE}],Projection={ProjectionType=ALL},ProvisionedThroughput={ReadCapacityUnits=5,WriteCapacityUnits=5}'\
        --stream-specification \
            StreamEnabled=true,StreamViewType=NEW_AND_OLD_IMAGES
    aws dynamodb update-time-to-live \
        --endpoint-url=http://$HOST:4566 \
        --table-name table2 \
        --time-to-live-specification Enabled=true,AttributeName=_ttl
```

**Comment by whummer** [comment]
Thanks for reporting @mariaines . Can you please share some more details about your setup . Which code is running in the `ss2_gql_1` container, and how is it configured to interact with the LocalStack container? Also, you're using a fairly old image tag `0.12.4` - any chance we can give it a try with `latest` (after pulling the latest image)? If you can share some more details on how to reproduce (ideally a small Github repo with the full config), we'll be happy to look into it. Thanks!

**Comment by mariaines** [comment]
Hey @whummer apologies for the delay. Same problem with latest image. More details below

ss2_gql_1 container is running a uvicorn app serving a GraphQL endpoint. (In production this is a ApiGateway Lambda using mangum.) It interacts with localstack via the local_services_network (see full docker-compose below). 

We also have a dynamo stream Lambda, which in turn may write back to dynamo. Localstack is configured with 
```yml
      - LAMBDA_EXECUTOR=docker-reuse
      - LAMBDA_REMOTE_DOCKER=False
      - LAMBDA_DOCKER_NETWORK=local_services_network
```
We see the same timeout problem in the container running the stream Lambda.

In the logs above we are writing to dynamo via localstack like so:
```python
from contextlib import AsyncExitStack
import aiobotocore

session = aiobotocore.get_session()
_exit_stack = AsyncExitStack()
_db_client = None

async def _create_client(*args, **kwargs):
    client = session.create_client(*args, **kwargs)
    client = await _exit_stack.enter_async_context(client)
    return client

async def get_aws_db_client():
    global _db_client
    if _db_client is None:
        _db_client = await _create_client(
            "dynamodb",
            endpoint_url="http://localstack:4566",
            region_name="us-west-2",
        )
    return _db_client

async def save(self):
    dynamodb = await get_aws_db_client()
    params = {
                "TableName": "table2",
                "Key": ...,
                "UpdateExpression": ...,
                "ExpressionAttributeValues": ...,
                "ExpressionAttributeNames": ...,
                "ReturnValues": ...,
                "ReturnConsumedCapacity": "INDEXES",
            }
    await dynamodb.update_item(**params)
```

Full docker-compose.yml:
```yml
version: "3"

services:
  # Localstack service
  localstack:
    image: localstack/localstack-full:latest
    ports:
      - "4566:4566"  # Edge port
      - "${PORT_WEB_UI-8080}:${PORT_WEB_UI-8080}" # Console UI
      - "5891:5891"
    environment:
      - DOCKER_HOST=unix:///var/run/docker.sock
      - SERVICES=dynamodb,dynamodbstreams,kinesis,lambda,sqs
      - DEFAULT_REGION=us-west-2
      - DEBUG=1
      # Runs Lamba function in lambci container that matches
      # Lamba environment configuration
      - LAMBDA_EXECUTOR=docker-reuse
      - LAMBDA_REMOTE_DOCKER=False
      - LAMBDA_DOCKER_NETWORK=local_services_network
      # If true, will remove idle Lambda containers
      - LAMBDA_REMOVE_CONTAINERS=False
      - AWS_DEFAULT_REGION=us-west-2
      - AWS_CONTAINER_CREDENTIALS_RELATIVE_URI=/creds
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      # Mount the Lambda build directory so Lambdas can be shared across containers
      - /tmp/.lambda-build:/tmp/.lambda-build
    tmpfs:
      - /tmp/localstack:exec,mode=600
    networks:
      services_network:
        ipv4_address: "169.254.170.3"
    depends_on:
      - ecs-local-endpoints

  elasticsearch:
    # Use 6.2 to match what's currently running on stats ES cluster
    image: docker.elastic.co/elasticsearch/elasticsearch:6.2.2
    ports:
      - "4571:4571"
      - "9300:9300"
    environment:
      - bootstrap.memory_lock=true
      - discovery.type=single-node
      - "ES_JAVA_OPTS=-Xms200m -Xmx200m"
      - cluster.routing.allocation.disk.threshold_enabled=false
      - http.port=4571
    ulimits:
      memlock:
        soft: -1
        hard: -1
    user: 'elasticsearch'
    tmpfs:
      # Don't preserve state across container invocations
      - /usr/share/elasticsearch/data
    networks:
      services_network:
        ipv4_address: "169.254.170.4"

  # Build Lambdas and configure Localstack
  # Based upon https://docs.aws.amazon.com/lambda/latest/dg/python-package.html
  configure-localstack:
    image: lambci/lambda:build-python3.8
    volumes:
      - ..:/var/task:delegated
      - /tmp/.lambda-build:/tmp/.lambda-build
    environment:
      # Does not need real credentials to connect to Localstack
      - AWS_DEFAULT_REGION=us-west-2
      - AWS_ACCESS_KEY_ID=dummyaccess
      - AWS_SECRET_ACCESS_KEY=dummysecret
    entrypoint: /bin/sh -c
    command: "/var/task/local/configure-localstack.sh"
    networks:
      services_network:
        ipv4_address: "169.254.170.5"
    depends_on:
      - localstack

  # This container vends AWS credentials to other containers
  ecs-local-endpoints:
    # The Amazon ECS Local Container Endpoints Docker Image
    image: amazon/amazon-ecs-local-container-endpoints
    volumes:
      # Mount /var/run so we can access docker.sock and talk to Docker
      - /var/run:/var/run
      # Mount the shared configuration directory, used by the AWS CLI and AWS SDKs
      - $HOME/.aws/:/home/.aws/
    environment:
      # define the home folder; credentials will be read from $HOME/.aws
      HOME: "/home"
      # You can change which AWS CLI Profile is used
      AWS_PROFILE: "default"
    networks:
      services_network:
        # This special IP address is recognized by the AWS SDKs and AWS CLI
        ipv4_address: "169.254.170.2"

  # GraphQL API/Websocket server
  ss2_gql:
    build:
      context: .
      dockerfile: Dockerfile
    ports:
      - 8000:8000
      - 5890:5890
    volumes:
      - ../../:/home/ec2-user
    environment:
      # Used to determine if we're local, latest, or prod
      - ENV=local
      # Used to obtain DynamoDB prod credentials via ecs-local-endpoints
      - AWS_DEFAULT_REGION=us-west-2
      - AWS_CONTAINER_CREDENTIALS_RELATIVE_URI=/creds
      # Enable debugging
      - DEBUG=True
      # Debug listening port
      - DEBUG_PORT=5890
      # General log verbosity
      - SS_LOG_VERBOSITY=2
      # Model log verbosity
      - DY_LOG_VERBOSITY=2
      # Use LocalStack to emulate Dynamo
      - LOCALSTACK_HOSTNAME=localstack
      - SS2_GQL=True
    networks:
      services_network:
        ipv4_address: "169.254.170.6"
    depends_on:
      - localstack
      - configure-localstack
      - ecs-local-endpoints

networks:
  # This special network is configured so that the local metadata
  # service can bind to the specific IP address that ECS uses
  # in production
  services_network:
    driver: bridge
    ipam:
      config:
        - subnet: "169.254.170.0/24"
        #  gateway: 169.254.170.1
```

configure-localstack.sh issues a series of aws cli commands to localstack to create our dynamo table, lambda functions, sqs queues, etc, e.g.
```sh
    aws dynamodb create-table \
        --endpoint-url=http://$HOST:4566 \
        --table-name table2 \
        --attribute-definitions \
            AttributeName=_pk,AttributeType=S \
            AttributeName=_sk,AttributeType=S \
            AttributeName=_gsi1_pk,AttributeType=S \
            AttributeName=_gsi1_sk,AttributeType=S \
            AttributeName=_gsi2_pk,AttributeType=S \
            AttributeName=_gsi2_sk,AttributeType=S \
        --key-schema \
            AttributeName=_pk,KeyType=HASH \
            AttributeName=_sk,KeyType=RANGE \
        --provisioned-throughput \
            ReadCapacityUnits=5,WriteCapacityUnits=5 \
        --global-secondary-indexes \
            'IndexName=_gsi1_pk-_gsi1_sk-index,KeySchema=[{AttributeName=_gsi1_pk,KeyType=HASH},{AttributeName=_gsi1_sk,KeyType=RANGE}],Projection={ProjectionType=ALL},ProvisionedThroughput={ReadCapacityUnits=5,WriteCapacityUnits=5}'\
            'IndexName=_gsi2_pk-_gsi2_sk-index,KeySchema=[{AttributeName=_gsi2_pk,KeyType=HASH},{AttributeName=_gsi2_sk,KeyType=RANGE}],Projection={ProjectionType=ALL},ProvisionedThroughput={ReadCapacityUnits=5,WriteCapacityUnits=5}'\
        --stream-specification \
            StreamEnabled=true,StreamViewType=NEW_AND_OLD_IMAGES
  aws lambda create-function \
    --endpoint-url http://$HOST:4566 \
    --function-name DynamoDbStreamFunction \
    --runtime python3.8 \
    --memory-size 1024 \
    --role arn:aws:iam:isignored \
    --handler db_stream.stream_handler \
    --code S3Bucket="__local__",S3Key="$BUILDDIR" \
    --timeout 600 \
    --dead-letter-config TargetArn=arn:aws:sqs:us-west-2:000000000000:DynamoDbStreamDeadLetterQueue \
    --environment Variables="{ENV=local,DB_STREAM_ERROR_QUEUE=DynamoDbStreamErrorQueue,NOTIFY_PERSON_QUEUE=NotifyPersonQueue}"\
    --publish
  aws lambda create-event-source-mapping \
    --endpoint-url http://$HOST:4566 \
    --function-name DynamoDbStreamFunction \
    --batch-size 10 --starting-position LATEST \
    --event-source-arn arn:aws:dynamodb:us-west-2:000000000000:table/table2 \
    --starting-position LATEST
```

**Comment by whummer** [comment]
Thanks for the update @mariaines . Definitely looks like this could be a networking issue inside the Docker compose stack. You mentioned that you have configured `LAMBDA_DOCKER_NETWORK=local_services_network`, but in your config I can see `services_network` under the `networks` section. Is `local_services_network` referring to the actual name of the docker-compose network you're trying to attach to - can you please share the output of this:
```
$ docker network ls | grep services_network
```

**Comment by mariaines** [comment]
Hmm, good catch.. yes the network is present as local_services_network:
```
$ docker network ls | grep services_network
0116b98fa75e   local_services_network   bridge    local
```
"local" is the folder name where docker-compose.yml lives, maybe that is setting the project name?

**Comment by mariaines** [comment]
Hi @whummer, I was able to set up a lil demo repo: https://github.com/mariaines/localstack-dynamo-timeout

Please take a look and let me know if you need further details.

**Comment by mariaines** [comment]
Update: I asked a teammate to try to repro and they are seeing a slightly different error, but with similar symptoms. When they trigger the /create_item endpoint, the write to dynamo from uvicorn_app goes through, as evidenced by the item appearing in the table and the DynamoDbStreamHandler being triggered, BUT the `update_item` call hangs in uvicorn_app and eventually times out. When the stream handler running in a Localstack-spun-up container attempts to write back to dynamo, it gets `NoCredentialsError` from boto. 

This suggests to me that there's some conflict between the way localstack handles AWS credentials and the way we are configuring the ecs-local-endpoints container to vend real credentials to other containers. (We are doing this because we need to talk to some real AWS services outside of localstack.) Perhaps some low-level boto code is getting confused? Also weird that my teammate is seeing the `NoCredentialsError` while I just see the same timeout in the stream listener...

```sh
localstack_1            | [ERROR] NoCredentialsError: Unable to locate credentials
localstack_1            | Traceback (most recent call last):
localstack_1            |   File "/var/task/db_stream.py", line 10, in stream_handler
localstack_1            |     asyncio.get_event_loop().run_until_complete(
localstack_1            |   File "/var/lang/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
localstack_1            |     return future.result()
localstack_1            |   File "/var/task/db_stream.py", line 16, in _event_handler
localstack_1            |     await _process_record(record, lambda_context)
localstack_1            |   File "/var/task/db_stream.py", line 29, in _process_record
localstack_1            |     result = await dynamodb.update_item(
localstack_1            |   File "/var/task/aiobotocore/client.py", line 140, in _make_api_call
localstack_1            |     http, parsed_response = await self._make_request(
localstack_1            |   File "/var/task/aiobotocore/client.py", line 160, in _make_request
localstack_1            |     return await self._endpoint.make_request(operation_model, request_dict)
localstack_1            |   File "/var/task/aiobotocore/endpoint.py", line 85, in _send_request
localstack_1            |     request = await self.create_request(request_dict, operation_model)
localstack_1            |   File "/var/task/aiobotocore/endpoint.py", line 78, in create_request
localstack_1            |     await self._event_emitter.emit(event_name, request=request,
localstack_1            |   File "/var/task/aiobotocore/hooks.py", line 27, in _emit
localstack_1            |     response = await handler(**kwargs)
localstack_1            |   File "/var/task/aiobotocore/signers.py", line 16, in handler
localstack_1            |     return await self.sign(operation_name, request)
localstack_1            |   File "/var/task/aiobotocore/signers.py", line 63, in sign
localstack_1            |     auth.add_auth(request)
localstack_1            |   File "/var/task/botocore/auth.py", line 357, in add_auth
localstack_1            |     raise NoCredentialsError
```

**Comment by mariaines** [comment]
I should mention that the ecs-local-endpoints container expects real AWS credentials to live in `$HOME/.aws/` on the host

**Comment by mariaines** [comment]
Update: reverting to Localstack 0.12.3 seems to fix this socket timeout issue! (But we need other fixes available in 0.12.4)

Hope that helps y'all pinpoint the issue.

**Comment by whummer** [comment]
Thanks for the update @mariaines . It seems that we have been able to find the root cause of this issue. A fix will be pushed in PR  #3765 shortly.

Meanwhile, can you please try enabling the following configuration in your LocalStack container environment:
```
SYNCHRONOUS_DYNAMODB_EVENTS=false
```

Please keep us posted how it goes. Thanks for your help!

**Comment by mariaines** [comment]
Hi @whummer, thanks for the update. I've added this to my config, and so far, so good! Thank you so much!
